Bottom: a28aed160b4bda1fe569fc7a51fd77088efa2c68
Top:    d72f7db240fa874512ae9ef17fcb18a2808e7116
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-11-06 01:32:12 +0000

Try a new, different sized VBO buffer for each upload.


---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index cdc01ce..5ed9400 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -135,8 +135,8 @@ hidgl_init_triangle_array (triangle_buffer *buffer)
   buffer->use_map = false;
 
   /* If using VBOs (but not mapping), we only need to this once */
-  if (buffer->use_vbo && !buffer->use_map)
-    glBufferData (GL_ARRAY_BUFFER, BUFFER_SIZE, NULL, GL_STREAM_DRAW);
+//  if (buffer->use_vbo && !buffer->use_map)
+//    glBufferData (GL_ARRAY_BUFFER, BUFFER_SIZE, NULL, GL_STREAM_DRAW);
 
   buffer->triangle_array = NULL;
   hidgl_reset_triangle_array (buffer);
@@ -176,9 +176,15 @@ hidgl_flush_triangles (triangle_buffer *buffer)
       buffer->triangle_array = NULL;
     } else {
       /* NB: We only upload the portion of the buffer we've used */
+#if 0
       glBufferSubData (GL_ARRAY_BUFFER, 0,
                        BUFFER_STRIDE * buffer->vertex_count,
                        buffer->triangle_array);
+#endif
+      glBufferData (GL_ARRAY_BUFFER,
+                    BUFFER_STRIDE * buffer->vertex_count,
+                    buffer->triangle_array,
+                    GL_STREAM_DRAW);
     }
   } else {
     data_pointer = buffer->triangle_array;
diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index d9420e7..79619aa 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -1045,6 +1045,8 @@ ghid_end_drawing (GHidPort *port)
   else
     glFlush ();
 
+  gdk_gl_drawable_wait_gl (pGlDrawable);
+
   /* end drawing to current GL-context */
   gdk_gl_drawable_gl_end (pGlDrawable);
 }
